### PR TITLE
Added JaxPrinter

### PR DESCRIPTION
**Pull request by tttc3**
#### References to other Issues or PRs
Adds JAX ~~and parametric printing support~~ as discussed in #20516 
Parametric support has been moved to #23661.

#### Brief description of what is fixed or changed
Allows `lambdify` to output JAX compatible functions. ~~and provides parametric support for all children of `AbstractPythonCodePrinter`.~~

~~The parametric option attempts to provide native support for the problems that @patrick-kidger's `sympytorch` and @milescranmer `sympy2jax` packages solve. I hope this implementation achieves this/will achive this if #20516 is merged?~~

#### Release Notes

* printing
  * Added `JaxPrinter`, allowing `sympy.lambdify` to operate on and produce JAX functions. Example: `f = lambdify(variables, expr, 'jax')`.

### CODE REVIEW

File: sympy/printing/tests/test_numpy.py [hunk 1]
```diff
@@ -210,49 +214,49 @@ def test_expm1():
...
+    assert abs(f(1e-10) - 1e-10 - 5e-21) <= NUMPY_DEFAULT_EPSILON
 
 
 def test_log1p():
     if not np:
         skip("NumPy not installed")
 
     f = lambdify((a,), log1p(a), 'numpy')
-    assert abs(f(1e-99) - 1e-99) < 1e-100
+    assert abs(f(1e-99) - 1e-99) <= NUMPY_DEFAULT_SMALLEST_NORMAL
```
Comment (bjodah) [inline sympy/printing/tests/test_numpy.py:225]: this should probably be based on 1e-99 (e.g. epsilon*1e-99) or we should check that the difference is exactly zero.

File: sympy/utilities/lambdify.py [hunk 2]
```diff
@@ -346,6 +350,24 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
...
+        useful in scenarios where these numeric values need to be
+        optimised/modified, but manually changing these values and regenerating
+        the function would be inefficient/undesirable.
+
+        When ``True``, the function returned by lambdify will take the below
+        modified form, where params are the default parameter values used in the
+        generated function:
+
+        >>> f, params = lambdify(x, 2*x + 1, parametric=True)
+        >>> f(1)
```
Comment (bjodah) [inline sympy/utilities/lambdify.py:364]: Maybe we should show the contents of `params` too?

**Review by bjodah — COMMENTED** [review]
Thank you for this contribution, I had a quick look and it looks really nice.

The Jax printer looks good, the `parametric` part I'm less sure about: isn't this approach fragile to e.g. reordering of the arguments? What is the benefit over simply preprocessing the expression and introducing parameter symbols manually?

Would it perhaps be possible to split the PR into two parts, first the Jax printer, and then a second PR for the parametric part?

Let's see what the others say.

**Comment by moorepants** [comment]
I also don't understand the parametric part. If you internally replace all numeric nodes with symbols, how would the user know how to provide values in the resulting numeric function in order and consistently. For example, what if there are two 1's in the expression? Are you replacing those with the same symbol or two different symbols and, if two, what if the user then provides different values for them? Maybe a description of the use case for parametric would help clear it up.

**Comment by moorepants** [comment]
> Would it perhaps be possible to split the PR into two parts, first the Jax printer, and then a second PR for the parametric part?

This sounds like a good idea.

File: sympy/printing/tests/test_numpy.py (hunk 1 above)
Comment (tttc3) [reply sympy/printing/tests/test_numpy.py]: I have updated this and `test_expm1` such that the epsilon is correctly scaled

**Comment by oscarbenjamin** [comment]
Is the parametric part in a separate PR now?

Now that the parametric part has been removed from this PR can someone review this?

It looks like a good addition for 1.11.

**Comment by oscarbenjamin** [comment]
Also there's a simple merge conflict.

**Comment by tttc3** [comment]
> Is the parametric part in a separate PR now?
> 
> Now that the parametric part has been removed from this PR can someone review this?
> 
> It looks like a good addition for 1.11.

The parameteric pull has been moved to #23661

**Review by bjodah — COMMENTED** [review]
I made a second pass of reviewing. Some minor comments/questions.

File: sympy/printing/numpy.py [hunk 3]
```diff
@@ -445,3 +446,43 @@ def __init__(self, settings=None):
...
+    _kf = _jax_known_functions
+    _kc = _jax_known_constants
+
+    def __init__(self, settings=None):
+        super().__init__(settings=settings)
+
+    # These need specific override to allow for the lack of "jax.numpy.reduce"
+    def _print_And(self, expr):
+        "Logical And printer"
+        return "{}({}.asarray([{}]), axis=0)".format(
```
[the remark this answers is not in the document]
Comment (tttc3) [reply sympy/printing/numpy.py:470]: I can add tests for `_print_And` and `_print_Or`. The `axis=0` ensures the functionality is the same as the numpy printer. Given a condition such as `e = (x >= 1) & (x < 2)`, `JaxPrinter` will generate `jax.numpy.all(jax.numpy.asarray([jax.numpy.greater_equal(x, 1), jax.numpy.less(x, 2)]), axis=0)`. If `x` is an array such as `[0,1,2]` then the output of the inner part of the function (the part that is passed to `all`) will be `[[False, True, True], [True, True, False]]`. To get the correct result (`[False, True, False]`), we then need to reduce the results along axis=0, which is performed with `all(, axis=0)`. For the `_print_Or` case `all(, axis=0)` is replaced with `any(, axis=0)`.

File: bin/test_optional_dependencies.py [hunk 4]
```diff
@@ -34,6 +34,10 @@ class TestsFailedError(Exception):
     # aesara
     '*aesara*',
 
+    # jax
+    '*jax*',
+    'sympy/utilities/tests/test_lambdify.py',
```
Comment (asmeurer) [inline bin/test_optional_dependencies.py:39]: test_lambdify is already in the list above.

File: sympy/printing/tests/test_jax.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    f = lambdify(M, M + Identity(2))
+    ma = jax.numpy.array([[1, 2], [3, 4]])
+    mr = jax.numpy.array([[2, 2], [3, 5]])
+
+    assert (f(ma) == mr).all()
+
+    from sympy.core.symbol import symbols
+    n = symbols('n', integer=True)
+    N = MatrixSymbol("M", n, n)
+    raises(NotImplementedError, lambda: lambdify(N, N + Identity(n)))
```
[the remark this answers is not in the document]
Comment (tttc3) [reply sympy/printing/tests/test_jax.py]: Updated in commit 4b3869b

File: sympy/printing/numpy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tttc3) [reply sympy/printing/numpy.py:470]: These tests have been added in 4b3869b. See lines 190->199 in `test_jax.py`.

[the remark this answers is not in the document]
Comment (bjodah) [reply sympy/printing/numpy.py:470]: I see. Great, thanks!

**Review by bjodah — APPROVED** [review]
This looks ready from my point of view.
Thank you @tttc3 for this contribution.

**Comment by patrick-kidger** [comment]
Folks over here may be interested in [sympy2jax](https://github.com/google/sympy2jax), which also aims to do the sympy->JAX conversion (and back again). I think the main differences to over here are (a) parameter handling and (b) there's no printing; it's just a normal JAX construct.

**Comment by asmeurer** [comment]
Thanks @patrick-kidger. Would it make sense to include sympy2jax in SymPy itself? Would you say it's strictly better than what's included here? I can imagine benefits to the printing approach (like being easier to debug), but I suspect sympy2jax approach is better in most cases.

**Comment by patrick-kidger** [comment]
IMO it is strictly better (I made it for a reason after all) but obviously I'm not an unbiased observer here.

I'm don't think it makes sense to bake into SymPy itself. It make an opionated choice of library to build on: Equinox as opposed to Flax or Haiku. And longer-term I don't think SymPy should need to implement a SymPy-to-Foo for those Foo that have yet to be released. Delegating to to some external library has worked so far [for PyTorch](https://github.com/patrick-kidger/sympytorch).
